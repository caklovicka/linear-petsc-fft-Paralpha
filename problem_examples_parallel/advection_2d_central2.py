import numpy as np
from scipy import sparse
from petsc4py import PETSc
from core_parallel.linear_paralpha import LinearParalpha

"""
advection eq. in 2d
u_t + c_x * u_x + c_y * u_y = f
"""


class Advection(LinearParalpha):

    # user defined, just for this class
    c = [1, 1]
    X_left = 0
    X_right = 1
    Y_left = 0
    Y_right = 1
    xx = None
    yy = None

    # PROCESSORS NEEDED = proc_row * proc_col

    def __init__(self):

        LinearParalpha.__init__(self)

    def setup(self):

        # ---- PRESETUP ----

        self.xx = np.linspace(self.X_left, self.X_right, self.spatial_points[0] + 2)[1:-1]
        self.yy = np.linspace(self.Y_left, self.Y_right, self.spatial_points[1] + 2)[1:-1]

        self.dx = []
        self.dx.append((self.X_right - self.X_left) / (self.spatial_points[0] + 1))
        self.dx.append((self.Y_right - self.Y_left) / (self.spatial_points[1] + 1))

        # x and size_global_A have to be filled before super().setup()
        self.x = np.meshgrid(self.xx, self.yy)
        self.global_size_A = 1
        for n in self.spatial_points:
            self.global_size_A *= n

        # ---- PRESETUP <end> ----

        # do not delete this, this builds communicators needed for your matrix Apar
        super().setup()

        # ---- POSTSETUP ----

        # user defined, parallel matrix, rows are divided between processors in group 'comm_matrix'
        # rows_beg and rows_end define the chunk of matrix Apar in this communicator

        # user defined matrix Apar, sparse form recommended
        row = list()
        col = list()
        data = list()
        for i in range(self.row_beg, self.row_end, 1):
            # right x diagonal
            if (i+1) % self.spatial_points[0] != 0:
                row.append(i)
                col.append(i+1)
                data.append(-self.c[0] / (2 * self.dx[0]))
            # left x diagonal
            if i % self.spatial_points[0] != 0:
                row.append(i)
                col.append(i - 1)
                data.append(self.c[0] / (2 * self.dx[0]))

            # left y diagonal
            if i >= self.spatial_points[0]:
                row.append(i)
                col.append(i-self.spatial_points[0])
                data.append(self.c[1] / (2 * self.dx[1]))
            # right y diagonal
            if i + self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i+self.spatial_points[0])
                data.append(-self.c[1] / (2 * self.dx[1]))

        data = np.array(data)
        row = np.array(row) - self.row_beg
        col = np.array(col)
        self.Apar = sparse.csr_matrix((data, (row, col)), shape=(self.row_end - self.row_beg, self.global_size_A))

        # ---- POSTSETUP <end> ----

    # user defined
    def bpar(self, t):

        bb = self.rhs(t, self.x).flatten()
        bb[:self.spatial_points[0]] += self.c[1] / (2 * self.dx[1]) * self.e(t, self.xx)
        bb[-self.spatial_points[0]:] -= self.c[1] / (2 * self.dx[1]) * self.d(t, self.xx)
        bb[::self.spatial_points[0]] += self.c[0] / (2 * self.dx[0]) * self.a(t, self.yy)
        bb[self.spatial_points[0]-1::self.spatial_points[0]] -= self.c[0] / (2 * self.dx[0]) * self.b(t, self.yy)

        return bb[self.row_beg:self.row_end]

    # user defined
    def u_exact(self, t, z):
        return np.sin(2 * np.pi * (z[0] - self.c[0] * t)) * np.sin(2 * np.pi * (z[1] - self.c[1] * t))

    # user defined
    def a(self, t, y):
        return self.u_exact(t, (self.X_left, y))

    # user defined
    def b(self, t, y):
        return self.u_exact(t, (self.X_right, y))

    # user defined
    def e(self, t, x):
        return self.u_exact(t, (x, self.Y_left))

    # user defined
    def d(self, t, x):
        return self.u_exact(t, (x, self.Y_right))

    # user defined
    def u_initial(self, z):
        return self.u_exact(self.T_start, z)

    # user defined
    @staticmethod
    def rhs(t, z):
        return 0 * z[0] * z[1]

    @staticmethod
    def norm(x):
        return np.linalg.norm(x.real, np.inf)

    # petsc solver on comm_matrix
    def linear_solver(self, M_loc, m_loc, m0_loc, tol):
        m = PETSc.Vec()
        m.createWithArray(array=m_loc, comm=self.comm_matrix)
        m0 = PETSc.Vec()
        m0.createWithArray(array=m0_loc, comm=self.comm_matrix)
        M = PETSc.Mat()
        csr = (M_loc.indptr, M_loc.indices, M_loc.data)
        M.createAIJ(size=(self.global_size_A, self.global_size_A), csr=csr, comm=self.comm_matrix)

        ksp = PETSc.KSP()
        ksp.create()
        ksp.setType('gmres')
        ksp.setFromOptions()
        ksp.setTolerances(rtol=tol, atol=tol, max_it=self.global_size_A)
        pc = ksp.getPC()
        pc.setType('none')
        ksp.setOperators(M)
        ksp.setInitialGuessNonzero(True)
        ksp.solve(m, m0)
        sol = m0.getArray()
        it = ksp.getIterationNumber()

        m.destroy()
        m0.destroy()
        ksp.destroy()
        M.destroy()

        return sol, it







