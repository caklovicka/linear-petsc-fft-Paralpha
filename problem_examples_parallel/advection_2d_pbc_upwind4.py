import numpy as np
from scipy import sparse
from petsc4py import PETSc
from core_parallel.linear_paralpha import LinearParalpha

"""
advection1 eq. in 2d
u_t + c_x * u_x + c_y * u_y = f
"""


class Advection(LinearParalpha):

    # user defined, just for this class
    c = [1, 1]
    X_left = 0
    X_right = 1
    Y_left = 0
    Y_right = 1
    xx = None
    yy = None

    # PROCESSORS NEEDED = proc_row * proc_col

    def __init__(self):

        LinearParalpha.__init__(self)

    def setup(self):

        # ---- PRESETUP ----

        self.xx = np.linspace(self.X_left, self.X_right, self.spatial_points[0] + 1)[:-1]
        self.yy = np.linspace(self.Y_left, self.Y_right, self.spatial_points[1] + 1)[:-1]

        self.dx = []
        self.dx.append((self.X_right - self.X_left) / self.spatial_points[0])
        self.dx.append((self.Y_right - self.Y_left) / self.spatial_points[1])

        # x and size_global_A have to be filled before super().setup()
        self.x = np.meshgrid(self.xx, self.yy)
        self.global_size_A = 1
        for n in self.spatial_points:
            self.global_size_A *= n

        # ---- PRESETUP <end> ----

        # do not delete this, this builds communicators needed for your matrix Apar
        super().setup()

        # ---- POSTSETUP ----

        # user defined, parallel matrix, rows are divided between processors in group 'comm_matrix'
        # rows_beg and rows_end define the chunk of matrix Apar in this communicator

        # user defined matrix Apar, sparse form recommended
        row = list()
        col = list()
        data = list()
        cx = -self.c[0] / (60 * self.dx[0])
        cy = -self.c[1] / (60 * self.dx[1])
        for i in range(self.row_beg, self.row_end, 1):

            row.append(i)
            col.append((i - 3) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(-5 * cx)

            row.append(i)
            col.append((i - 2) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(30 * cx)

            row.append(i)
            col.append((i - 1) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(-90 * cx)

            row.append(i)
            col.append(i % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(50 * cx + 50 * cy)

            row.append(i)
            col.append((i + 1) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(15 * cx)

            row.append(i)
            col.append((i - 3 * self.spatial_points[0]) % self.global_size_A)
            data.append(-5 * cy)

            row.append(i)
            col.append((i - 2 * self.spatial_points[0]) % self.global_size_A)
            data.append(30 * cy)

            row.append(i)
            col.append((i - self.spatial_points[0]) % self.global_size_A)
            data.append(-90 * cy)

            row.append(i)
            col.append((i + self.spatial_points[0]) % self.global_size_A)
            data.append(15 * cy)

        data = np.array(data)
        row = np.array(row) - self.row_beg
        col = np.array(col)
        self.Apar = sparse.csr_matrix((data, (row, col)), shape=(self.row_end - self.row_beg, self.global_size_A))
        del data, row, col

        # ---- POSTSETUP <end> ----

    # user defined
    def bpar(self, t):
        return self.rhs(t, self.x).flatten()[self.row_beg:self.row_end]

    # user defined
    def u_exact(self, t, z):
        return np.sin(2 * np.pi * (z[0] - self.c[0] * t)) * np.sin(2 * np.pi * (z[1] - self.c[1] * t))

    # user defined
    def u_initial(self, z):
        return self.u_exact(self.T_start, z)

    # user defined
    @staticmethod
    def rhs(t, z):
        return 0 * z[0] * z[1]

    @staticmethod
    def norm(x):
        return np.linalg.norm(x.real, np.inf)

    # petsc solver on comm_matrix
    def linear_solver(self, M_loc, m_loc, m0_loc, tol):
        m = PETSc.Vec()
        m.createWithArray(array=m_loc, comm=self.comm_matrix)
        m0 = PETSc.Vec()
        m0.createWithArray(array=m0_loc, comm=self.comm_matrix)
        M = PETSc.Mat()
        csr = (M_loc.indptr, M_loc.indices, M_loc.data)
        M.createAIJ(size=(self.global_size_A, self.global_size_A), csr=csr, comm=self.comm_matrix)

        ksp = PETSc.KSP()
        ksp.create()
        ksp.setType('gmres')
        ksp.setFromOptions()
        ksp.setTolerances(rtol=tol, max_it=self.smaxiter)
        pc = ksp.getPC()
        pc.setType('none')
        ksp.setOperators(M)
        ksp.setInitialGuessNonzero(True)
        ksp.solve(m, m0)
        sol = m0.getArray()
        it = ksp.getIterationNumber()

        m.destroy()
        m0.destroy()
        ksp.destroy()
        M.destroy()

        return sol, it







