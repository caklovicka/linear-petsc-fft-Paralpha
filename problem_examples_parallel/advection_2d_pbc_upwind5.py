import numpy as np
from scipy import sparse
from petsc4py import PETSc
from mpi4py import MPI
from core_parallel.linear_paralpha import LinearParalpha

"""
advection1 eq. in 2d
u_t + c_x * u_x + c_y * u_y = f
"""


class Advection(LinearParalpha):

    # user defined, just for this class
    c = [1, 1]
    X_left = 0
    X_right = 1
    Y_left = 0
    Y_right = 1
    xx = None
    yy = None

    # PROCESSORS NEEDED = proc_row * proc_col

    def __init__(self):

        LinearParalpha.__init__(self)

    def setup(self):

        # ---- PRESETUP ----

        self.xx = np.linspace(self.X_left, self.X_right, self.spatial_points[0] + 1)[:-1]
        self.yy = np.linspace(self.Y_left, self.Y_right, self.spatial_points[1] + 1)[:-1]

        self.dx = []
        self.dx.append((self.X_right - self.X_left) / self.spatial_points[0])
        self.dx.append((self.Y_right - self.Y_left) / self.spatial_points[1])

        # x and size_global_A have to be filled before super().setup()
        self.x = np.meshgrid(self.xx, self.yy)
        self.global_size_A = 1
        for n in self.spatial_points:
            self.global_size_A *= n

        # ---- PRESETUP <end> ----

        # do not delete this, this builds communicators needed for your matrix Apar
        super().setup()

        # ---- POSTSETUP ----

        # user defined, parallel matrix, rows are divided between processors in group 'comm_matrix'
        # rows_beg and rows_end define the chunk of matrix Apar in this communicator

        # user defined matrix Apar, sparse form recommended
        row = list()
        col = list()
        data = list()
        cx = -self.c[0] / (60 * self.dx[0])
        cy = -self.c[1] / (60 * self.dx[1])
        for i in range(self.row_beg, self.row_end, 1):
            row.append(i)
            col.append((i - 4) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(3 * cx)

            row.append(i)
            col.append((i - 3) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(-20 * cx)

            row.append(i)
            col.append((i - 2) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(60 * cx)

            row.append(i)
            col.append((i - 1) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(-120 * cx)

            row.append(i)
            col.append(i % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(65 * cx + 65 * cy)

            row.append(i)
            col.append((i + 1) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(12 * cx)

            row.append(i)
            col.append((i - 4 * self.spatial_points[0]) % self.global_size_A)
            data.append(3 * cy)

            row.append(i)
            col.append((i - 3 * self.spatial_points[0]) % self.global_size_A)
            data.append(-20 * cy)

            row.append(i)
            col.append((i - 2 * self.spatial_points[0]) % self.global_size_A)
            data.append(60 * cy)

            row.append(i)
            col.append((i - self.spatial_points[0]) % self.global_size_A)
            data.append(-120 * cy)

            row.append(i)
            col.append((i + self.spatial_points[0]) % self.global_size_A)
            data.append(12 * cy)

        data = np.array(data)
        row = np.array(row) - self.row_beg
        col = np.array(col)
        self.Apar = sparse.csr_matrix((data, (row, col)), shape=(self.row_end - self.row_beg, self.global_size_A))

        # ---- POSTSETUP <end> ----
        return

    # user defined
    def bpar(self, t):
        return self.rhs(t, self.x).flatten()[self.row_beg:self.row_end]

    # user defined
    def u_exact(self, t, z):
        return np.sin(2 * np.pi * (z[0] - self.c[0] * t)) * np.sin(2 * np.pi * (z[1] - self.c[1] * t))

    # user defined
    def u_initial(self, z):
        return self.u_exact(self.T_start, z)

    # user defined
    @staticmethod
    def rhs(t, z):
        return 0 * z[0] * z[1]

    @staticmethod
    def norm(x):
        return np.linalg.norm(x.real, np.inf)

    # petsc solver on comm_matrix
    def linear_solver(self, M_loc, m_loc, m0_loc, tol):
        print('tu sam')
        m = PETSc.Vec()
        m.createWithArray(array=m_loc, comm=self.comm_matrix)
        m0 = PETSc.Vec()
        m0.createWithArray(array=m0_loc, comm=self.comm_matrix)
        M = PETSc.Mat()
        csr = (M_loc.indptr, M_loc.indices, M_loc.data)
        M.createAIJ(size=(self.global_size_A, self.global_size_A), csr=csr, comm=self.comm_matrix)

        ksp = PETSc.KSP()
        ksp.create(comm=self.comm_matrix)
        ksp.setType('gmres')
        ksp.setFromOptions()
        ksp.setTolerances(rtol=tol, max_it=self.smaxiter)
        pc = ksp.getPC()
        pc.setType('none')
        ksp.setOperators(M)
        ksp.setInitialGuessNonzero(True)
        ksp.solve(m, m0)
        sol = m0.getArray()
        it = ksp.getIterationNumber()

        m.destroy()
        m0.destroy()
        ksp.destroy()
        M.destroy()

        return sol, it