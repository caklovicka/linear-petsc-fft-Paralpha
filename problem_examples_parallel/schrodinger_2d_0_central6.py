import numpy as np
from scipy import sparse
from core_parallel.linear_paralpha import LinearParalpha
import scipy as sc
from petsc4py import PETSc

"""
schrodinger in 2d, 2nd order in space
u_t = c ( u_xx + u_yy )
"""


class Schrodinger(LinearParalpha):

    # user defined, just for this class
    c = 1j
    sigma = 1
    p = [1, 1]
    X_left = -10
    X_right = 10
    Y_left = -10
    Y_right = 10
    xx = None
    yy = None

    # PROCESSORS NEEDED = proc_row * proc_col

    def __init__(self):

        LinearParalpha.__init__(self)

    def setup(self):

        # ---- PRESETUP ----

        self.xx = np.linspace(self.X_left, self.X_right, self.spatial_points[0] + 2)[1:-1]
        self.yy = np.linspace(self.Y_left, self.Y_right, self.spatial_points[1] + 2)[1:-1]

        self.dx = []
        self.dx.append((self.X_right - self.X_left) / (self.spatial_points[0] + 1))
        self.dx.append((self.Y_right - self.Y_left) / (self.spatial_points[1] + 1))

        # x and size_global_A have to be filled before super().setup()

        self.x = np.meshgrid(self.xx, self.yy)
        del self.xx, self.yy

        self.global_size_A = 1
        for n in self.spatial_points:
            self.global_size_A *= n

        # ---- PRESETUP <end> ----

        # do not delete this, this builds communicators needed for your matrix Apar
        super().setup()

        # ---- POSTSETUP ----

        # user defined, parallel matrix, rows are divided between processors in group 'comm_matrix'
        # rows_beg and rows_end define the chunk of matrix Apar in this communicator

        # user defined matrix Apar, sparse form recommended
        row = list()
        col = list()
        data = list()
        cx = self.c / self.dx[0]**2
        cy = self.c / self.dx[1]**2
        for i in range(self.row_beg, self.row_end, 1):
            # x part
            # i - 3
            if i % self.spatial_points[0] - 3 >= 0:
                row.append(i)
                col.append(i - 3)
                data.append(cx / 90)

            # i - 2
            if i % self.spatial_points[0] - 2 >= 0:
                row.append(i)
                col.append(i - 2)
                data.append(-3 * cx / 20)

            # i - 1
            if i % self.spatial_points[0] - 1 >= 0:
                row.append(i)
                col.append(i - 1)
                data.append(3 * cx / 2)

            # i
            row.append(i)
            col.append(i)
            data.append(-49 * cx / 18 - 49 * cy / 18)

            # i + 1
            if i % self.spatial_points[0] + 1 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 1)
                data.append(3 * cx / 2)

            # i + 2
            if i % self.spatial_points[0] + 2 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 2)
                data.append(-3 * cx / 20)

            # i + 3
            if i % self.spatial_points[0] + 3 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 3)
                data.append(cx / 90)

            # y part
            # i - 3
            if i - 3 * self.spatial_points[0] >= 0:
                row.append(i)
                col.append(i - 3 * self.spatial_points[0])
                data.append(cy / 90)

            # i - 2
            if i - 2 * self.spatial_points[0] >= 0:
                row.append(i)
                col.append(i - 2 * self.spatial_points[0])
                data.append(-3 * cy / 20)

            # i - 1
            if i - self.spatial_points[0] >= 0:
                row.append(i)
                col.append(i - self.spatial_points[0])
                data.append(3 * cy / 2)

            # i + 1
            if i + self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + self.spatial_points[0])
                data.append(3 * cy / 2)

            # i + 2
            if i + 2 * self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + 2 * self.spatial_points[0])
                data.append(-3 * cy / 20)

            # i + 3
            if i + 3 * self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + 3 * self.spatial_points[0])
                data.append(cy / 90)

        data = np.array(data)
        row = np.array(row) - self.row_beg
        col = np.array(col)
        self.Apar = sparse.csr_matrix((data, (row, col)), shape=(self.row_end - self.row_beg, self.global_size_A))

        del data, row, col

        # ---- POSTSETUP <end> ----

    # user defined
    def bpar(self, t):
        return self.rhs(t, self.x).flatten()[self.row_beg:self.row_end]

    # user defined
    def u_exact(self, t, z):
        a = self.sigma + 2j * t
        const = self.sigma / a * np.exp(self.sigma/2 * (self.p[0]**2 + self.p[1]**2) * (self.sigma/a - 1))
        exp1 = np.exp(1j * self.sigma / a * (self.p[0] * z[0] + self.p[1] * z[1]))
        exp2 = np.exp(-1/(2 * a) * (z[0]**2 + z[1]**2))
        return const * exp1 * exp2

    # user defined
    def u_initial(self, z):
        return self.u_exact(self.T_start, z)

    # user defined
    def rhs(self, t, z):
        return 0 * z[0] * z[1]

    @staticmethod
    def norm(x):
        return np.linalg.norm(x, np.inf)

    # petsc solver on comm_matrix
    def linear_solver(self, M_loc, m_loc, m0_loc, tol):
        m = PETSc.Vec()
        m.createWithArray(array=m_loc, comm=self.comm_matrix)
        m0 = PETSc.Vec()
        m0.createWithArray(array=m0_loc, comm=self.comm_matrix)
        M = PETSc.Mat()
        csr = (M_loc.indptr, M_loc.indices, M_loc.data)
        M.createAIJ(size=(self.global_size_A, self.global_size_A), csr=csr, comm=self.comm_matrix)

        ksp = PETSc.KSP()
        ksp.create()
        ksp.setType('gmres')
        ksp.setFromOptions()
        ksp.setTolerances(rtol=tol, max_it=self.smaxiter)
        pc = ksp.getPC()
        pc.setType('none')
        ksp.setOperators(M)
        ksp.setInitialGuessNonzero(True)
        ksp.solve(m, m0)
        sol = m0.getArray()
        it = ksp.getIterationNumber()

        m.destroy()
        m0.destroy()
        ksp.destroy()
        M.destroy()

        return sol, it
