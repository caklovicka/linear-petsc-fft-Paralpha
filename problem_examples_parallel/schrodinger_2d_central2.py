import numpy as np
from scipy import sparse
from core_parallel.linear_paralpha import LinearParalpha
import scipy as sc
from petsc4py import PETSc

"""
schrodinger in 2d, 2nd order in space
u_t = c ( u_xx + u_yy )
"""


class Schrodinger(LinearParalpha):

    # user defined, just for this class
    c = 1j
    sigma = 1
    p = [1, 1]
    X_left = -50
    X_right = 50
    Y_left = -50
    Y_right = 50
    xx = None
    yy = None

    # PROCESSORS NEEDED = proc_row * proc_col

    def __init__(self):

        LinearParalpha.__init__(self)

    def setup(self):

        # ---- PRESETUP ----

        self.xx = np.linspace(self.X_left, self.X_right, self.spatial_points[0] + 2)[1:-1]
        self.yy = np.linspace(self.Y_left, self.Y_right, self.spatial_points[1] + 2)[1:-1]

        self.dx = []
        self.dx.append((self.X_right - self.X_left) / (self.spatial_points[0] + 1))
        self.dx.append((self.Y_right - self.Y_left) / (self.spatial_points[1] + 1))

        # x and size_global_A have to be filled before super().setup()

        self.x = np.meshgrid(self.xx, self.yy)

        self.global_size_A = 1
        for n in self.spatial_points:
            self.global_size_A *= n

        # ---- PRESETUP <end> ----

        # do not delete this, this builds communicators needed for your matrix Apar
        super().setup()

        # ---- POSTSETUP ----

        # user defined, parallel matrix, rows are divided between processors in group 'comm_matrix'
        # rows_beg and rows_end define the chunk of matrix Apar in this communicator

        # user defined matrix Apar, sparse form recommended
        row = list()
        col = list()
        data = list()
        for i in range(self.row_beg, self.row_end, 1):

            row.append(i)
            col.append(i)
            data.append(-2 * self.c * (1/self.dx[0]**2 + 1/self.dx[1]**2))

            if (i + 1) % self.spatial_points[0] != 0:
                row.append(i)
                col.append(i + 1)
                data.append(self.c / self.dx[0] ** 2)

            if i % self.spatial_points[0] != 0 and 0 <= i - 1:
                row.append(i)
                col.append(i - 1)
                data.append(self.c / self.dx[0] ** 2)

            if i + self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + self.spatial_points[0])
                data.append(self.c / self.dx[1] ** 2)

            if 0 <= i - self.spatial_points[0]:
                row.append(i)
                col.append(i - self.spatial_points[0])
                data.append(self.c / self.dx[1] ** 2)

        data = np.array(data)
        row = np.array(row) - self.row_beg
        col = np.array(col)
        self.Apar = sparse.csr_matrix((data, (row, col)), shape=(self.row_end - self.row_beg, self.global_size_A))

        # ---- POSTSETUP <end> ----

    # user defined
    def bpar(self, t):
        cx = self.c / self.dx[0]**2
        cy = self.c / self.dx[1]**2
        bb = np.zeros(self.global_size_A, dtype=complex)
        bb += self.rhs(t, self.x).flatten()
        bb[::self.spatial_points[0]] += cx * self.a(t, self.xx)
        bb[self.spatial_points[0] - 1::self.spatial_points[0]] += cx * self.b(t, self.xx)
        bb[:self.spatial_points[0]] += cy * self.e(t, self.yy)
        bb[-self.spatial_points[0]:] += cy * self.d(t, self.yy)
        return bb[self.row_beg:self.row_end]

    # user defined
    def u_exact(self, t, z):
        a = self.sigma + 2j * t
        const = self.sigma / a * np.exp(self.sigma/2 * (self.p[0]**2 + self.p[1]**2) * (self.sigma/a - 1))
        exp1 = np.exp(1j * self.sigma / a * (self.p[0] * z[0] + self.p[1] * z[1]))
        exp2 = np.exp(-1/(2 * a) * (z[0]**2 + z[1]**2))
        return const * exp1 * exp2

    # user defined
    def a(self, t, y):
        return self.u_exact(t, (self.X_left, y))

    # user defined
    def b(self, t, y):
        return self.u_exact(t, (self.X_right, y))

    # user defined
    def e(self, t, x):
        return self.u_exact(t, (x, self.Y_left))

    # user defined
    def d(self, t, x):
        return self.u_exact(t, (x, self.Y_right))

    # user defined
    def u_initial(self, z):
        return self.u_exact(self.T_start, z)

    # user defined
    def rhs(self, t, z):
        return 0 * z[0] * z[1]

    @staticmethod
    def norm(x):
        return np.linalg.norm(x.real, np.inf)

    # petsc solver on comm_matrix
    def linear_solver(self, M_loc, m_loc, m0_loc, tol):
        m = PETSc.Vec()
        m.createWithArray(array=m_loc, comm=self.comm_matrix)
        m0 = PETSc.Vec()
        m0.createWithArray(array=m0_loc, comm=self.comm_matrix)
        M = PETSc.Mat()
        csr = (M_loc.indptr, M_loc.indices, M_loc.data)
        M.createAIJ(size=(self.global_size_A, self.global_size_A), csr=csr, comm=self.comm_matrix)

        ksp = PETSc.KSP()
        ksp.create()
        ksp.setType('gmres')
        ksp.setFromOptions()
        ksp.setTolerances(rtol=tol, atol=tol, max_it=self.global_size_A)
        pc = ksp.getPC()
        pc.setType('none')
        ksp.setOperators(M)
        ksp.setInitialGuessNonzero(True)
        ksp.solve(m, m0)
        sol = m0.getArray()

        m.destroy()
        m0.destroy()
        ksp.destroy()
        M.destroy()

        return sol
