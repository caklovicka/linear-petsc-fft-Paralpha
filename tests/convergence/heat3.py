# the following lines disable the numpy multithreading [optional]
import os
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["OPENBLAS_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"
os.environ["VECLIB_MAXIMUM_THREADS"] = "1"
os.environ["NUMEXPR_NUM_THREADS"] = "1"

import sys
sys.path.append('../..')
import numpy as np

from problem_examples_parallel.heat_2d_pbc_central6 import Heat
prob = Heat()

# choosing a number of points
prob.spatial_points = [300, 300]
prob.time_points = 3

# choosing a time domain
prob.T_start = np.pi
prob.T_end = prob.T_start + 0.1

# choosing the number of intervals handled in parallel
prob.time_intervals = 16
prob.rolling = 1

# choosing a parallelization strategy
prob.proc_col = 1
prob.proc_row = prob.time_intervals

# choosing a solver
prob.solver = 'custom'

# setting maximum number of iterations
prob.maxiter = 3
prob.smaxiter = 1000

# choosing a setting for the alpha sequence
prob.optimal_alphas = False
prob.alphas = [1e-12]

# setting tolerances
prob.tol = 1e-12
prob.stol = 1e-13

prob.setup()                                # must be before solve()
prob.solve()                                # this is where magic happens
prob.summary(details=True)
