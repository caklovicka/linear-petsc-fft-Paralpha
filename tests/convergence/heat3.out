on 15,  abs, rel err inf norm [from paralpha] = 1.8553292250129867e-10, 1.864644681657757e-10, iter = 0, rolling = 0
on 15,  abs, rel err inf norm [from paralpha] = 4.802824944587035e-13, 4.826939536725478e-13, iter = 1, rolling = 0
-----------------------< summary >-----------------------
solving on T1 = 3.141592653589793, T2 = 3.241592653589793
no. of spatial points = [300, 300]
dx = [0.0033333333333333335, 0.0033333333333333335]
no. of time points on an interval = 3
no. of time intervals = 16
no. of alphas = 1, [0.0001]
rolling intervals = 1
dt (of one SDC interval) = 0.0062500000000000056
processors for spatial parallelization for solving (I - Q x A) are 1
processors for time interval parallelization are 16
maxiter = 3
output document = None
tol = 1e-12
last error = NotImplemented
residuals = [[0.0029191083278541363, 4.99583758362121e-07, 7.083222897108499e-13]]
consecutive errors= [[inf, 0.0049958349075071105, 4.6280767471892105e-07]]
iterations of Paralpha = [2.]
max iterations of Paralpha = 2.0
algorithm time = 27.49495 s
communication time = 0.07267 s
inner solver = custom
system_time_max = [10.131696551999994, 17.16376938899998]
system_time_min = [8.839058987000044, 14.157360427000185]
solver_its_max = [[360, 540]]
solver_its_min = [[232, 280]]
inner solver tols = [1e-13, 1e-13]
inner solver maxiter = 1000
-----------------------< end summary >-----------------------
on 15,  abs, rel err inf norm [from paralpha] = 7.39653967066683e-11, 7.43367709279898e-11, iter = 0, rolling = 0
on 15,  abs, rel err inf norm [from paralpha] = 4.760636401300593e-13, 4.784539168205761e-13, iter = 1, rolling = 0
-----------------------< summary >-----------------------
solving on T1 = 3.141592653589793, T2 = 3.241592653589793
no. of spatial points = [300, 300]
dx = [0.0033333333333333335, 0.0033333333333333335]
no. of time points on an interval = 3
no. of time intervals = 16
no. of alphas = 1, [1e-08]
rolling intervals = 1
dt (of one SDC interval) = 0.0062500000000000056
processors for spatial parallelization for solving (I - Q x A) are 1
processors for time interval parallelization are 16
maxiter = 3
output document = None
tol = 1e-12
last error = NotImplemented
residuals = [[0.0029191083278541363, 3.0608049428337836e-10, 7.462919171530302e-13]]
consecutive errors= [[inf, 0.004995834664143552, 7.072216117391003e-11]]
iterations of Paralpha = [2.]
max iterations of Paralpha = 2.0
algorithm time = 35.88793 s
communication time = 0.07325 s
inner solver = custom
system_time_max = [9.810102080999968, 25.876640467000016]
system_time_min = [8.953460698999834, 22.906883917000187]
solver_its_max = [[330, 990]]
solver_its_min = [[237, 560]]
inner solver tols = [1e-13, 1e-13]
inner solver maxiter = 1000
-----------------------< end summary >-----------------------
on 15,  abs, rel err inf norm [from paralpha] = 1.2172806246272216e-07, 1.2233924913139302e-07, iter = 0, rolling = 0
on 15,  abs, rel err inf norm [from paralpha] = 3.5129678401591842e-12, 3.5306061650280476e-12, iter = 1, rolling = 0
on 15,  abs, rel err inf norm [from paralpha] = 3.5380587348760923e-12, 3.5558230390799237e-12, iter = 2, rolling = 0
-----------------------< summary >-----------------------
solving on T1 = 3.141592653589793, T2 = 3.241592653589793
no. of spatial points = [300, 300]
dx = [0.0033333333333333335, 0.0033333333333333335]
no. of time points on an interval = 3
no. of time intervals = 16
no. of alphas = 1, [1e-12]
rolling intervals = 1
dt (of one SDC interval) = 0.0062500000000000056
processors for spatial parallelization for solving (I - Q x A) are 1
processors for time interval parallelization are 16
maxiter = 3
output document = None
tol = 1e-12
last error = NotImplemented
residuals = [[0.0029191083278541363, 7.923060121317604e-07, 2.55253596037619e-11]]
consecutive errors= [[inf, 0.0049959134218909304, 8.553689399940269e-08, 6.452967568344924e-12]]
iterations of Paralpha = [3.]
max iterations of Paralpha = 3.0
algorithm time = 67.47630 s
communication time = 0.10106 s
inner solver = custom
system_time_max = [9.342440421999981, 28.734131985999966, 29.144131378999873]
system_time_min = [8.80663671100001, 27.954408150000063, 28.536642347999987]
solver_its_max = [[300, 1000, 1000]]
solver_its_min = [[245, 795, 826]]
inner solver tols = [1e-13, 1e-13, 1e-13]
inner solver maxiter = 1000
-----------------------< end summary >-----------------------
